trustfall_core/src/ir/ty.rs

```diff
@@ -0,0 +1,332 @@
+        self.modifiers.nullable()
+    }
+
+    /// Returns whether the type is a list or not.
+    ///
+    /// # Example
+    /// ```
+    /// use trustfall_core::ir::ty::Type;
+    ///
+    /// let non_null_int_arr = Type::new("[Int!]").unwrap();
+    /// assert_eq!(non_null_int_arr.is_list(), true);
+    ///
+    /// let non_null_int = Type::new("Int!").unwrap();
+    /// assert_eq!(non_null_int.is_list(), false);
+    /// ```
+    pub fn is_list(&self) -> bool {
+        self.modifiers.is_list()
+    }
+
+    /// Returns the type inside the outermost list of this type if it is a list, otherwise returns `None`.
+    ///
+    /// # Example
+    /// ```
+    /// use trustfall_core::ir::ty::Type;
+    ///
+    /// let non_null_int_arr = Type::new("[Int!]").unwrap();
+    /// let non_null_int = Type::new("Int!").unwrap();
+    /// assert_eq!(non_null_int_arr.as_list(), Some(non_null_int.clone()));
+    /// assert_eq!(non_null_int.as_list(), None);
+    /// ```
+    pub fn as_list(&self) -> Option<Self> {
+        Some(Self { base: Arc::clone(&self.base), modifiers: self.modifiers.as_list()? })
+    }
+
+    /// Returns the type of the elements of the first individual type found inside this type.
+    ///
+    /// # Example
+    /// ```
+    /// use trustfall_core::ir::ty::Type;
+    ///
+    /// let int_list_ty = Type::new("[Int!]").unwrap();
+    /// assert_eq!(int_list_ty.base_named_type(), "Int");
+    ///
+    /// let string_ty = Type::new("String!").unwrap();
+    /// assert_eq!(string_ty.base_named_type(), "String");
+    pub fn base_named_type(&self) -> &str {
+        &self.base
+    }
+
+    /// Convert a [`GQLType`] to a [`Type`].
+    pub(crate) fn from_type(ty: &GQLType) -> Type {
+        let mut base = &ty.base;
+
+        let mut mask = if ty.nullable { 0 } else { Modifiers::NON_NULLABLE_MASK };
+
+        let mut i = 0;
+
+        while let BaseType::List(ty_inside_list) = base {
+            mask |= Modifiers::LIST_MASK << i;
+            i += 2;
+            if i > Modifiers::MAX_LIST_DEPTH * 2 {
+                panic!("too many nested lists");
+            }
+            if !ty_inside_list.nullable {
+                mask |= Modifiers::NON_NULLABLE_MASK << i;
+            }
+            base = &ty_inside_list.base;
+        }
+
+        let BaseType::Named(name) = base else {
+            unreachable!(
+                "should be impossible to get a non-named type after looping through all list types"
+            )
+        };
+        Type { base: name.to_string().into(), modifiers: Modifiers { mask } }
+    }
+}
+
+impl Display for Type {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        // left
+        {
+            let mut current = Some(self.modifiers.clone());
+            while let Some(mods) = current {
+                if mods.is_list() {
+                    write!(f, "[")?;
+                }
+                current = mods.as_list();
+            }
+        }
```

Comment:
Perhaps indicative of a missing helper method for counting the number of nested lists?

Should be just a mask and a popcount.